google/datalab/ml/_metrics.py

```diff
@@ -0,0 +1,268 @@
+      schema_file: Path to a JSON file containing BigQuery schema. Used if "headers" is None.
+
+    Returns:
+      a Metrics instance.
+
+    Raises:
+      ValueError if both headers and schema_file are None.
+    """
+
+    if headers is not None:
+      names = headers
+    elif schema_file is not None:
+      with _util.open_local_or_gcs(schema_file, mode='r') as f:
+        schema = json.load(f)
+      names = [x['name'] for x in schema]
+    else:
+      raise ValueError('Either headers or schema_file is needed')
+
+    metrics = Metrics(input_csv_pattern=input_csv_pattern, headers=names)
+    return metrics
+
+  @staticmethod
+  def from_bigquery(sql):
+    """Create a Metrics instance from a bigquery query or table.
+
+    Returns:
+      a Metrics instance.
+
+    Args:
+      bigquery: A BigQuery table name or a query.
+    """
+
+    if isinstance(sql, bq.Query):
+      sql = sql._expanded_sql()
+
+    parts = sql.split('.')
+    if len(parts) == 1 or len(parts) > 3 or any(' ' in x for x in parts):
+      sql = '(' + sql + ')'  # query, not a table name
+    else:
+      sql = '`' + sql + '`'  # table name
+
+    metrics = Metrics(bigquery=sql)
+    return metrics
+
+  def _get_data_from_csv_files(self):
+    """Get data from input csv files."""
+
+    all_df = []
+    for file_name in self._input_csv_files:
+      with _util.open_local_or_gcs(file_name, mode='r') as f:
+        all_df.append(pd.read_csv(f, names=self._headers))
+    df = pd.concat(all_df, ignore_index=True)
+    return df
+
+  def _get_data_from_bigquery(self, queries):
+    """Get data from bigquery table or query."""
+
+    all_df = []
+    for query in queries:
+      all_df.append(query.execute().result().to_dataframe())
+    df = pd.concat(all_df, ignore_index=True)
+    return df
+
+  def accuracy(self):
+    """Get accuracy numbers for each target and overall.
+
+    Returns:
+      A DataFrame with two columns: 'class' and 'accuracy'. It contains the overall
+      accuracy with class being '_all'.
+
+    Raises:
+      Exception if the CSV headers do not include 'target' or 'predicted', or BigQuery
+      does not return 'target' or 'predicted' column.
+    """
+
+    if self._input_csv_files:
+      df = self._get_data_from_csv_files()
+      if 'target' not in df or 'predicted' not in df:
+        raise ValueError('Cannot find "target" or "predicted" column')
+
+      labels = sorted(set(df['target']) | set(df['predicted']))
+      accuracy_results = []
+
+      for label in labels:
+        correct_count = len(df[(df['target'] == df['predicted']) & (df['target'] == label)])
+        total_count = len(df[(df['target'] == label)])
+        accuracy_results.append({
+            'target': label,
+            'accuracy': float(correct_count) / total_count,
+            'count': total_count
+        })
+
+      total_correct_count = len(df[(df['target'] == df['predicted'])])
+      total_accuracy = float(total_correct_count) / len(df)
```

Comment:
Just double checking, is len(df) ever zero?

Reply:
Done. Also did similar check in a few other places, and tested 0 instance in csv and 0 instance in bq.